File: Python/MergeSort.py
def Merge(arr, left, middle, right):
    LeftHalf = middle - left + 1
    RightHalf = right - middle
    LeftArr = []
    RightArr = []
    for i in range(LeftHalf):
        LeftArr.append(arr[left + i])
    for i in range(RightHalf):
        RightArr.append(arr[middle + i + 1])
    i = 0
    j = 0
    k = left
    while i < LeftHalf and j < RightHalf:
        if LeftArr[i] <= RightArr[j]:
            arr[k] = LeftArr[i]
            i = i + 1
        else:
            arr[k] = RightArr[j]
            j = j + 1
        k = k + 1
    while i < LeftHalf:
        arr[k] = LeftArr[i]
        i = i + 1
        k = k + 1
    while j < RightHalf:
        arr[k] = RightArr[j]
        j = j + 1
        k = k + 1

def MergeSort(arr, left, right):
    if left < right:
        middle = left + (right - left) // 2
        MergeSort(arr, left, middle)
        MergeSort(arr, middle + 1, right)
        Merge(arr, left, middle, right)
try:
    arr = []
    size = int(input("Enter the size of the list: "))
    for _ in range(size):
        while True:
            try:
                el = int(input("Enter a number: "))
                arr.append(el)
                break 
            except ValueError:
                print("Invalid input. Please enter a valid number.")

    print("INPUT LIST:", arr)
    MergeSort(arr, 0, size - 1)
    print("SORTED LIST:", arr)
except KeyboardInterrupt:
    print("Program interrupted by the user.")
except Exception as e:
    print(f"INPUT MUST BE IN NUMBER")

File: Java/MergeSort.java
import java.util.ArrayList;
import java.util.Scanner;

public class MergeSort {
    public static void main(String args[]) {
        ArrayList<Integer> arr = new ArrayList<>();
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter size of the list: ");
        int size = scanner.nextInt();
        for (int i = 0; i < size; i++) {
            int el = scanner.nextInt();
            arr.add(el);
        }
        System.out.println("INPUT LIST");
        for (int num : arr)
            System.out.print(num + " ");
        mergesort(arr);
        System.out.println("\nSORTED LIST");
        for (int num : arr)
            System.out.print(num + " ");
        scanner.close();
    }

    public static void mergesort(ArrayList<Integer> arr) {
        if (arr.size() > 1) {
            int middle = arr.size() / 2;
            ArrayList<Integer> LeftHalf = new ArrayList<>(arr.subList(0, middle));
            ArrayList<Integer> RightHalf = new ArrayList<>(arr.subList(middle, arr.size() ));
            mergesort(LeftHalf);
            mergesort(RightHalf);
            ActualWorker(arr, LeftHalf, RightHalf);
        }
    }

    public static void ActualWorker(ArrayList<Integer> arr, ArrayList<Integer> left, ArrayList<Integer> right) {
        int i = 0, j = 0, k = 0;
        while (i < left.size() && j < right.size()) {
            if (left.get(i) <= right.get(j)) {
                arr.set(k, left.get(i));
                i++;
            } else {
                arr.set(k, right.get(j));
                j++;
            }
            k++;
        }

        while (i < left.size()) {
            arr.set(k, left.get(i));
            i++;
            k++;
        }

        while (j < right.size()) {
            arr.set(k, right.get(j));
            j++;
            k++;
        }
    }
}
